wallet-new/test/unit/UTxO/BlockGen.hs

```diff
@@ -0,0 +1,245 @@
+runBlockGenWith
+    :: Hash h Addr
+    => (BlockGenCtx h -> BlockGenCtx h)
+    -> Transaction h Addr
+    -> BlockGen h a
+    -> Gen a
+runBlockGenWith settings boot (BlockGen m) =
+    evalStateT m (settings (initializeCtx boot))
+
+-- | Create an initial context from the boot transaction.
+initializeCtx :: Hash h Addr => Transaction h Addr -> BlockGenCtx h
+initializeCtx boot@Transaction{..} = BlockGenCtx {..}
+  where
+    _bgcAddressesToBalances =
+        foldl' outToValue mempty trOuts
+    outToValue acc (Output {..}) =
+        Map.insertWith (+) outAddr outVal acc
+    _bgcFreshHash = 0
+    _bgcCurrentBlockchain = ledgerSingleton boot
+
+-- | Lift a 'Gen' action into the 'BlockGen' monad.
+liftGen :: Gen a -> BlockGen h a
+liftGen = BlockGen . lift
+
+-- | Provide a fresh hash value for a transaction.
+freshHash :: BlockGen h Int
+freshHash = do
+    i <- use bgcFreshHash
+    bgcFreshHash += 1
+    pure i
+
+-- | Select a random address from the current state.
+selectFromAddr :: BlockGen h Addr
+selectFromAddr =
+    liftGen . elements =<< uses bgcAddressesToBalances (toList . Map.keysSet)
+
+-- | Select a random address from the current state that is not the
+-- provided address.
+selectToAddress :: Addr -> BlockGen h Addr
+selectToAddress addr = do
+    addrs <- uses bgcAddressesToBalances (toList . Set.delete addr . Map.keysSet)
+    liftGen (elements (filter (not . isAvvmAddr) addrs))
+
+-- | Returns true if the 'addrActorIx' is the 'IxAvvm' constructor.
+isAvvmAddr :: Addr -> Bool
+isAvvmAddr addr =
+    case addrActorIx addr of
+        IxAvvm _ -> True
+        _ -> False
+
+-- | Select a random address from the current set of addresses, and
+-- a random value between 0 and that address's total value minus the
+-- possible fee amount. The 'Value' that is returned is not deducted from
+-- the 'Addr's balance.
+selectFromAddrAndAmount :: BlockGen h (Addr, Value)
+selectFromAddrAndAmount = do
+    addrs <- uses bgcAddressesToBalances Map.toList
+    (addr, amount) <- liftGen $ elements addrs
+    value <- liftGen $ choose (1, amount `safeSubtract` maxFee)
+    if value < 1
+        then localBlockGen
+            (bgcAddressesToBalances . at addr .~ Nothing)
+            selectFromAddrAndAmount
+        else pure (addr, value)
+
+maxFee :: Num a => a
+maxFee = 180000
+
+-- | Run a 'BlockGen' action with a modified context. Changes to the state
+-- from the inner action are discarded.
+localBlockGen
+    :: (BlockGenCtx h -> BlockGenCtx h)
+    -> BlockGen h a
+    -> BlockGen h a
+localBlockGen f action = do
+    st <- get
+    modify f
+    r <- action
+    put st
+    pure r
+
+-- | 'Value' is an alias for 'Word64', which underflows. This detects
+-- underflow and returns @0@ for underflowing values.
+safeSubtract :: Value -> Value -> Value
+safeSubtract x y
+    | z > x     = 0
+    | otherwise = z
+  where
+    z = x - y
+
+-- | Create a fresh transaction that depends on the fee provided to it.
```

Comment:
This function is the meat of the generator. The other generators just repeat this one some random amount of times.